types/deep-freeze/index.d.ts

```diff
@@ -1,13 +1,13 @@
-// Type definitions for deep-freeze
+// Type definitions for deep-freeze 0.1
 // Project: https://github.com/substack/deep-freeze
 // Definitions by: Bart van der Schoor <https://github.com/Bartvds>
 // Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+// TypeScript Version: 2.1
 
-declare var deepFreeze: DeepFreeze.DeepFreezeInterface;
 export = deepFreeze;
 
-declare namespace DeepFreeze {
-   export interface DeepFreezeInterface {
-      <T>(obj: T): T;
-    }
-}
+declare function deepFreeze<T>(a: T[]): ReadonlyArray<DeepReadonly<T>>;
+declare function deepFreeze<T extends Function>(f: T): T;
+declare function deepFreeze<T>(o: T): DeepReadonly<T>;
+
+type DeepReadonly<T> = Readonly<{ [P in keyof T]: DeepReadonly<T[P]> }>;
```

Comment:
One problem is that consumers can no longer reference `DeepReadonly`, so ideally you'd do function/namespace merging, export `DeepReadonly` from the namespace, and `export =` the declarations. So


```ts
declare function deepFreeze<T>(a: T[]): ReadonlyArray<deepFreeze.DeepReadonly<T>>;
declare function deepFreeze<T extends Function>(f: T): T;
declare function deepFreeze<T>(o: T): deepFreeze.DeepReadonly<T>;

declare namespace deepFreeze {
    export type DeepReadonly<T> = Readonly<{ [P in keyof T]: DeepReadonly<T[P]> }>;
}

export deepFreeze
```